File: src/listener.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"



void chatterCallback(const std_msgs::String::ConstPtr& msg) // 定义回调函数，参数是一个指向接收到的消息的智能指针
{
    ROS_INFO("I heard: [%s]", msg->data.c_str()); // 输出日志信息，显示接收到的消息内容
}

int main(int argc, char **argv) // ROS节点的入口函数
{
    ros::init(argc, argv, "listener"); // 初始化ROS节点，节点名称为listener
    ros::NodeHandle n; // 创建一个NodeHandle对象，负责与ROS系统进行通信
    ros::Subscriber sub = n.subscribe("chatter", 1000, chatterCallback); // 创建一个Subscriber对象，订阅名为"chatter"的主题，队列大小为1000，回调函数为chatterCallback

    ros::spin(); // 进入循环，等待回调函数被调用

    return 0;
}

File: scripts/listener.py
#!/usr/bin/env python3
import rospy
from std_msgs.msg import String


def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.data)  # 在ROS日志中输出接收到的消息内容

def listener():
    rospy.init_node('listener', anonymous=True)  # 初始化ROS节点，节点名称为'listener'，anonymous=True表示如果有同名节点会自动添加随机数后缀以区分
    rospy.Subscriber("chatter", String, callback)  # 创建一个订阅者对象，订阅名为'chatter'的话题，消息类型为String，回调函数为callback
    rospy.spin()  # 保持程序持续运行，等待回调函数被调用
if __name__ == '__main__':
    listener()  # 调用listener函数开始订阅消息  